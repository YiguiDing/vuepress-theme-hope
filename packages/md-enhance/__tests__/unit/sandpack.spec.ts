/* eslint-disable @typescript-eslint/naming-convention */
import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { mdSandpack } from "../../src/node/markdown-it/sandpack/index.js";

const decodeFiles = (content: string): Record<string, string> =>
  JSON.parse(
    Buffer.from(decodeURIComponent(content), "base64").toString(),
  ) as Record<string, string>;

const getTemplate = (renderResult: string): string | null => {
  const result = renderResult.match(/template="(.*?)"/s);

  if (!result) return null;

  return result[1];
};

const getFiles = (renderResult: string): Record<string, string> | null => {
  const result = renderResult.match(/files="(.*?)"/s);

  if (!result) return null;

  return decodeFiles(result[1]);
};

const getOptions = (renderResult: string): Record<string, unknown> | null => {
  const result = renderResult.match(/options="(.*?)"/s);

  if (!result) return null;

  return JSON.parse(decodeURIComponent(result[1])) as Record<string, unknown>;
};

const getCustomSetup = (
  renderResult: string,
): Record<string, unknown> | null => {
  const result = renderResult.match(/customSetup="(.*?)"/s);

  if (!result) return null;

  return JSON.parse(decodeURIComponent(result[1])) as Record<string, unknown>;
};

describe("Sandpack", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(mdSandpack);

  it("Should resolve sandpack info", () => {
    const result = markdownIt.render(
      `
::: sandpack#vue Sandpack demo

@file App.vue
\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const files = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("vue");

    expect(files).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
      "Comp.vue": `\
<template>
  <div>Comp</div>
</template>
`,
    });

    expect(options).toEqual({});
    expect(customSetup).toEqual({});
  });

  it("Should resolve sandpack info with options", () => {
    const result = markdownIt.render(
      `
::: sandpack Sandpack demo with options

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@options

\`\`\`js
{
  readOnly: true,
  showReadOnly: false
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("");

    expect(file).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
      "Comp.vue": `\
<template>
  <div>Comp</div>
</template>
`,
    });
    expect(options).toEqual({
      readOnly: true,
      showReadOnly: false,
    });

    expect(customSetup).toEqual({});
  });

  it("Should resolve sandpack info with customSetup", () => {
    const result = markdownIt.render(
      `
::: sandpack#vue Sandpack demo with customSetup

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@options

\`\`\`js
{
  readOnly: true,
  showReadOnly: false
}
\`\`\`

@setup

\`\`\`js
{
  entry: "/index.js",
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("vue");

    expect(file).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
    });

    expect(options).toEqual({
      readOnly: true,
      showReadOnly: false,
    });

    expect(customSetup).toEqual({
      entry: "/index.js",
    });
  });

  it("Should resolve sandpack info with file attrs", () => {
    const result = markdownIt.render(
      `
::: sandpack Sandpack demo with options

@file /src/App.vue {readOnly hidden}

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file /src/Comp.vue {active}

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@file Comp2.vue \\{hidden}

\`\`\`vue
<template>
  <div>Comp2</div>
</template>
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("");

    expect(file).toEqual({
      "/src/App.vue": {
        code: `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
        active: false,
        hidden: true,
        readOnly: true,
      },
      "/src/Comp.vue": {
        code: `\
<template>
  <div>Comp</div>
</template>
`,
        active: true,
        hidden: false,
        readOnly: false,
      },
      "Comp2.vue": `\
<template>
  <div>Comp2</div>
</template>
`,
    });
    expect(options).toEqual({});

    expect(customSetup).toEqual({});
  });
});
