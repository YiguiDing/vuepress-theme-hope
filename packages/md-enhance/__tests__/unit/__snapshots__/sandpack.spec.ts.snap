// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sandpack > Should resolve sandpack info 1`] = `
"<MdSandpack title=\\"Sandpack%20demo\\" key=\\"47bf4596\\" template=\\"vue\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4ifQ%3D%3D\\" options=\\"%7B%7D\\" customSetup=\\"%7B%7D\\">
</MdSandpack>
"
`;

exports[`Sandpack > Should resolve sandpack info with customSetup 1`] = `
"<MdSandpack title=\\"Sandpack%20demo%20with%20customSetup\\" key=\\"61e23a7c\\" template=\\"vue\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiJ9\\" options=\\"%7B%22readOnly%22%3Atrue%2C%22showReadOnly%22%3Afalse%7D\\" customSetup=\\"%7B%22entry%22%3A%22%2Findex.js%22%7D\\">
</MdSandpack>
"
`;

exports[`Sandpack > Should resolve sandpack info with file attrs 1`] = `
"<MdSandpack title=\\"Sandpack%20demo%20with%20options\\" key=\\"8b2108e4\\" template=\\"\\" files=\\"eyIvc3JjL0FwcC52dWUiOnsiY29kZSI6IjxzY3JpcHQgc2V0dXA%2BXG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCBtc2cgPSByZWYoJ0hlbGxvIFdvcmxkIScpXG48L3NjcmlwdD5cbjx0ZW1wbGF0ZT5cbiAgPGgxPnt7IG1zZyB9fTwvaDE%2BXG4gIDxpbnB1dCB2LW1vZGVsPVwibXNnXCIgLz5cbjwvdGVtcGxhdGU%2BXG4iLCJhY3RpdmUiOmZhbHNlLCJoaWRkZW4iOnRydWUsInJlYWRPbmx5Ijp0cnVlfSwiL3NyYy9Db21wLnZ1ZSI6eyJjb2RlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4iLCJhY3RpdmUiOnRydWUsImhpZGRlbiI6ZmFsc2UsInJlYWRPbmx5IjpmYWxzZX0sIkNvbXAyLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGRpdj5Db21wMjwvZGl2PlxuPC90ZW1wbGF0ZT5cbiJ9\\" options=\\"%7B%7D\\" customSetup=\\"%7B%7D\\">
</MdSandpack>
"
`;

exports[`Sandpack > Should resolve sandpack info with options 1`] = `
"<MdSandpack title=\\"Sandpack%20demo%20with%20options\\" key=\\"8b2108e4\\" template=\\"\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4ifQ%3D%3D\\" options=\\"%7B%22readOnly%22%3Atrue%2C%22showReadOnly%22%3Afalse%7D\\" customSetup=\\"%7B%7D\\">
</MdSandpack>
"
`;
